### PR TITLE
Replication Implementation

### PR DESCRIPTION
**Pull request by falfaroc**
# Description
Introduces the API hooks for PowerFlex replication calls and all tests associated with it to support the peer-to-peer replication feature.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/618 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Created all the necessary integration tests to ensure that PowerFlex replication is working according to the API calls.

Relevant integration test results.
```
...
=== RUN   TestGetPeerMDMs
    replication_test.go:91: PeerMDMID d02aebc400000000
--- PASS: TestGetPeerMDMs (0.06s)
=== RUN   TestGetTargetSystem
--- PASS: TestGetTargetSystem (0.04s)
=== RUN   TestGetProtectionDomain
--- PASS: TestGetProtectionDomain (0.21s)
=== RUN   TestTargetProtectionDomain
--- PASS: TestTargetProtectionDomain (0.03s)
=== RUN   TestStoragePools
--- PASS: TestStoragePools (0.17s)
=== RUN   TestLocateVolumesToBeReplicated
    replication_test.go:221: SourceVolume replication-source-5, TargetVolume replication-target-5
--- PASS: TestLocateVolumesToBeReplicated (0.07s)
=== RUN   TestCreateReplicationConsistencyGroup
--- PASS: TestCreateReplicationConsistencyGroup (5.05s)
=== RUN   TestGetReplicationConsistencyGroups
--- PASS: TestGetReplicationConsistencyGroups (0.04s)
=== RUN   TestAddReplicationPair
[FindVolumeID] volumeID: 6d68124c00000012
    replication_test.go:279: [TestAddReplicationPair] Local Volume ID: 6d68124c00000012
[FindVolumeID] volumeID: 847e0e2e00000005
    replication_test.go:287: [TestAddReplicationPair] Remote Volume ID: 847e0e2e00000005
    replication_test.go:303: ReplicationPairID: a491406200000000
--- PASS: TestAddReplicationPair (5.11s)
=== RUN   TestQueryReplicationPairs
--- PASS: TestQueryReplicationPairs (0.03s)
=== RUN   TestQueryReplicationPairsStatistics
    replication_test.go:336: Waiting for Replication Pair inttestrp to be complete.
    replication_test.go:341: Copied 0.000000
    replication_test.go:341: Copied 1.000000
    replication_test.go:348: Copy Complete: 1.000000
--- PASS: TestQueryReplicationPairsStatistics (10.12s)
=== RUN   TestCreateReplicationConsistencyGroupSnapshot
--- PASS: TestCreateReplicationConsistencyGroupSnapshot (0.03s)
=== RUN   TestSnapshotRetrieval
    replication_test.go:379: Remote Pair Volume: 847e0e2e00000005
    replication_test.go:397: Action Attributes Result: map[847e0e2e00000005:847e0e320000000c]
--- PASS: TestSnapshotRetrieval (0.05s)
=== RUN   TestExecuteFailoverOnReplicationGroup
--- PASS: TestExecuteFailoverOnReplicationGroup (0.04s)
=== RUN   TestExecuteRestoreOnReplicationGroup
    replication_test.go:624: Consistency Group is in Failover
--- PASS: TestExecuteRestoreOnReplicationGroup (2.12s)
=== RUN   TestExecuteSwitchoverOnReplicationGroup
    replication_test.go:607: Consistency Group d304296100000000 - Reached Consistency.
--- PASS: TestExecuteSwitchoverOnReplicationGroup (55.38s)
=== RUN   TestExecuteReverseOnReplicationGroup
    replication_test.go:624: Consistency Group is in Switchover
--- PASS: TestExecuteReverseOnReplicationGroup (6.22s)
=== RUN   TestExecutePauseOnReplicationGroup
    replication_test.go:607: Consistency Group d304296100000000 - Reached Consistency.
--- PASS: TestExecutePauseOnReplicationGroup (50.35s)
=== RUN   TestExecuteResumeOnReplicationGroup
--- PASS: TestExecuteResumeOnReplicationGroup (0.03s)
=== RUN   TestResizeReplicationPair
    replication_test.go:479: [TestResizeReplicationPair]  Failing over to get DR in Neutral state..
    replication_test.go:624: Consistency Group is in Failover
[FindVolumeID] volumeID: 6d68124c00000012
[FindVolumeID] volumeID: 847e0e2e00000005
    replication_test.go:624: Consistency Group is in Failover
    replication_test.go:607: Consistency Group d304296100000000 - Reached Consistency.
--- PASS: TestResizeReplicationPair (62.65s)
=== RUN   TestRemoveReplicationPairFromVolume
    replication_test.go:555: [TestRemoveReplicationPairFromVolume] Removed the following pair inttestrp
--- PASS: TestRemoveReplicationPairFromVolume (5.11s)
=== RUN   TestFreezeReplcationGroup
--- PASS: TestFreezeReplcationGroup (2.03s)
=== RUN   TestRemoveReplicationConsistencyGroup
--- PASS: TestRemoveReplicationConsistencyGroup (0.04s)
...
```

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -178,6 +178,11 @@ func (c *Client) getJSONWithRetry(
 	method, uri string,
 	body, resp interface{}) error {
 
+	if c == nil || c.api == nil {
```
Comment (VamsiSiddu-7) [inline api.go:181]: Do we really need this condition? if the client is nil. When we call this method from somewhere. There itself it panics with nil pointer dereference.

File: system.go [hunk 2]
```diff
@@ -69,6 +69,13 @@ func (c *Client) FindSystem(
 func (s *System) GetStatistics() (*types.Statistics, error) {
 	defer TimeSpent("GetStatistics", time.Now())
 
+	if s == nil || s.System == nil {
```
Comment (VamsiSiddu-7) [inline system.go:72]: Same comment here

**Review by VamsiSiddu-7 — COMMENTED** [review]
Can you please include the int-test result thanks.?

**Review by mbasha-dell — APPROVED** [review]

**Comment by falfaroc** [comment]
> Can you please include the int-test result thanks.?

Updated with relevant test results.

**Review by VamsiSiddu-7 — APPROVED** [review]

**Pull request merged**